### PR TITLE
Adjust to new tower auth mechanism

### PR DESCRIPTION
**Pull request by moolitayer**
Fix https://github.com/openshift/autoheal/issues/25

Tower introduced auth changes in a non backward compatible manner, see issue.
We will keep trying the previous `api/v2/authtoken` endpoint.
If that fails we will attmept the new `api/v2/users/<name>/personal_tokens`
If that fails too, we will error out.

### CODE REVIEW

**Comment by moolitayer** [comment]
@jhernand @zgalor @ironcladlou please take a look

File: pkg/awx/connection.go [hunk 1]
```diff
@@ -263,7 +305,7 @@ func (c *Connection) rawGet(path string, query url.Values) (output []byte, err e
 	if err != nil {
 		return
 	}
-	if response.StatusCode != 200 {
+	if response.StatusCode > 202 {
```
Comment (moolitayer) [inline pkg/awx/connection.go]: @jhernand this is needed since after creating the new token the server returns the `created` 201 status code.

File: pkg/awx/connection.go [hunk 2]
```diff
@@ -128,12 +136,14 @@ func (b *ConnectionBuilder) Build() (c *Connection, err error) {
...
+	authArgs := 0
+	for _, arg := range [3]string{b.username, b.token, b.bearer} {
+		if len(arg) > 0 {
+			authArgs++
+		}
 	}
-	if b.username == "" && b.token == "" {
-		err = fmt.Errorf("Either user name or token must be provided")
+	if authArgs != 1 {
+		err = fmt.Errorf("Exactly one of user name, token and bearer required")
```
Comment (zgalor) [inline pkg/awx/connection.go]: `Exactly one of the following is required: username, token or bearer`

File: pkg/awx/connection.go [hunk 3]
```diff
@@ -178,7 +188,7 @@ func (c *Connection) Close() {
 // new onw to the server.
 //
 func (c *Connection) ensureToken() error {
-	if c.token != "" {
+	if c.token != "" || c.bearer != "" {
```
Comment (zgalor) [inline pkg/awx/connection.go]: I sometimes see the test `len(x) != 0` and sometime `!= ""`.
Is there a difference? do we want to follow a standard?

File: pkg/awx/connection.go (hunk 1 above)
Comment (zgalor) [inline pkg/awx/connection.go]: don't we still want to check for other possible errors `!=200` ?
are codes 201-202 OK?
We might want to add a comment about that

File: pkg/awx/connection.go [hunk 4]
```diff
@@ -326,7 +368,7 @@ func (c *Connection) rawPost(path string, query url.Values, input []byte) (outpu
 	if err != nil {
 		return
 	}
-	if response.StatusCode != 200 {
+	if response.StatusCode > 202 {
```
Comment (zgalor) [inline pkg/awx/connection.go]: same as previous comment

File: pkg/awx/connection.go (hunk 3 above)
Comment (moolitayer) [reply pkg/awx/connection.go]: `len()` is more general (tests for nil as well) for string it's the same. I don't have a preference

File: pkg/awx/connection.go (hunk 1 above)
Comment (moolitayer) [reply pkg/awx/connection.go]: 201 Created (we create tokens in this PR), 202 Accepted is what you sometimes get when you launch a background job. These (and 200 of course)are the most common in my experience.

File: pkg/awx/connection.go (hunk 2 above)
Comment (moolitayer) [reply pkg/awx/connection.go]: Done

File: pkg/awx/connection.go (hunk 3 above)
Comment (moolitayer) [reply pkg/awx/connection.go]: I'll change the other occurrence in this PR to be != as well

**Comment by moolitayer** [comment]
Addressed comments @zgalor please take a look

**Comment by jhernand** [comment]
@moolitayer I would like to merge this, can you rebase?

**Comment by moolitayer** [comment]
@jhernand done
